### ISSUE TITLE
This package is (possibly) not async exception safe

### ISSUE THREAD

**Issue by parsonsmatt** [brandonchinn178/resourcet-pool#2]
... at least, I don't think it is. 

I'm debugging a `persistent` issue and by bypassing this library, I've fixed the bug. [This PR](https://github.com/yesodweb/persistent/pull/1207#issuecomment-800455760) links to the relevant issue with debugging and investigation history.

I know the code was vendored from the `persistent` library, but I think I'd caution folks against using it. The `unliftio-pool` package provides a non-`MonadBaseControl` interface to `resource-pool`.

**Comment by brandonchinn178** [comment]
Makes sense! I do notice that `mkAcquireType` uses `mask` from `Control.Exception`, which is probably why it doesn't handle async exceptions.

`unliftio-pool` is actually exactly what i needed. I can't believe I couldn't find it earlier.
